pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -389,6 +389,22 @@ func (configgen *ConfigGeneratorImpl) buildGatewayHTTPRouteConfig(node *model.Pr
 				}
 			}
 		}
+
+		// check all hostname in vHostDedupMap and if is not exist with HttpsRedirect set to true
+		// create VirtualHost to redirect
+		for _, hostname := range server.Hosts {
+			if _, exists := vHostDedupMap[host.Name(hostname)]; exists {
+				continue
+			} else if server.Tls != nil && server.Tls.HttpsRedirect {
+				newVHost := &route.VirtualHost{
+					Name:                       domainName(hostname, port),
+					Domains:                    buildGatewayVirtualHostDomains(hostname, port),
+					IncludeRequestAttemptCount: true,
+					RequireTls:                 route.VirtualHost_ALL,
+				}
+				vHostDedupMap[host.Name(hostname)] = newVHost
```

Comment:
It seems these are copied from L318-L333, so i guess we could remove the above `if len(virtualServices) == 0 && server.Tls != nil && server.Tls.HttpsRedirect {` branch

Reply:
yeah make sense, removing the duplicate code